be/src/exec/spill/spiller.hpp

```diff
@@ -63,72 +125,241 @@ Status Spiller::flush(RuntimeState* state, TaskExecutor&& executor, MemGuard&& g
     _running_flush_tasks++;
     // TODO: handle spill queue
     auto task = [this, state, guard = guard, mem_table = std::move(captured_mem_table)]() {
-        SCOPED_TIMER(_metrics.flush_timer);
+        RETURN_IF(!guard.scoped_begin(), Status::Cancelled("cancelled"));
+        SCOPED_TIMER(_spiller->metrics().flush_timer);
         DCHECK_GT(_running_flush_tasks, 0);
         DCHECK(has_pending_data());
-        guard.scoped_begin();
         //
         auto defer = DeferOp([&]() {
             {
                 std::lock_guard _(_mutex);
                 _mem_table_pool.emplace(std::move(mem_table));
             }
 
-            _update_spilled_task_status(_decrease_running_flush_tasks());
+            _spiller->update_spilled_task_status(_decrease_running_flush_tasks());
+            guard.scoped_end();
         });
 
-        _update_spilled_task_status(_run_flush_task(state, mem_table));
-        guard.scoped_end();
+        _spiller->update_spilled_task_status(flush_task(state, mem_table));
+        return Status::OK();
     };
     // submit io task
     RETURN_IF_ERROR(executor.submit(std::move(task)));
     return Status::OK();
 }
 
 template <class TaskExecutor, class MemGuard>
-StatusOr<ChunkPtr> Spiller::restore(RuntimeState* state, TaskExecutor&& executor, MemGuard&& guard) {
-    SCOPED_TIMER(_metrics.restore_timer);
-    DCHECK(has_output_data());
-
-    RETURN_IF_ERROR(_get_spilled_task_status());
-    RETURN_IF_ERROR(_acquire_input_stream(state));
-
-    ChunkPtr chunk;
-    // @TODO(silverbullet233): reuse ctx
-    SerdeContext ctx;
-    ASSIGN_OR_RETURN(chunk, _input_stream->get_next(ctx));
-    COUNTER_UPDATE(_metrics.restore_rows, chunk->num_rows());
-    _restore_read_rows += chunk->num_rows();
+StatusOr<ChunkPtr> SpillerReader::restore(RuntimeState* state, TaskExecutor&& executor, MemGuard&& guard) {
+    SCOPED_TIMER(_spiller->metrics().restore_timer);
+    ASSIGN_OR_RETURN(auto chunk, _current_stream->get_next(_spill_read_ctx));
     RETURN_IF_ERROR(trigger_restore(state, std::forward<TaskExecutor>(executor), std::forward<MemGuard>(guard)));
+    _read_rows += chunk->num_rows();
     return chunk;
 }
 
 template <class TaskExecutor, class MemGuard>
-Status Spiller::trigger_restore(RuntimeState* state, TaskExecutor&& executor, MemGuard&& guard) {
-    RETURN_IF_ERROR(_get_spilled_task_status());
-    RETURN_IF_ERROR(_acquire_input_stream(state));
-
-    DCHECK(_input_stream->enable_prefetch());
+Status SpillerReader::trigger_restore(RuntimeState* state, TaskExecutor&& executor, MemGuard&& guard) {
+    if (_current_stream == nullptr) {
+        return Status::OK();
+    }
+    DCHECK(_current_stream->enable_prefetch());
```

Comment:
In recent implementations, there will only be one stream.

Reply:
_current_stream indicates that there are man streams,  the streams is handled one-by-one, so can we use another  name, for an exmaple, just use _stream instead of _current_stream.

Reply:
fixed